persistent/persistence.py

```diff
@@ -315,7 +337,15 @@ def __setstate__(self, state):
                 raise TypeError('No instance dict')
             idict.clear()
             for k, v in inst_dict.items():
-                idict[intern(k)] = v
+                # Normally the keys for instance attributes are interned.
+                # Do that here, but only if it is possible to do so.
+
+                # TODO: On Python 2 codebases that straddle Python3,
+                # and use 'from __future__ import unicode_literals' it's not
+                # unheard of to have unicode objects in the __dict__ by accident.
+                # Should we watch for that and attempt to encode it so it can be
+                # interned?
```

Comment:
Unfortunately, we have objects in our production database that have unicode strings in their dicts because of exactly that reason. In our case, it was our own fault and we fixed the bug in our code, but that doesn't change the fact that the pickles exist. 

In general, it doesn't seem all that unlikely that using third-party libraries in a project could also introduce this problem.

Also, now that I look at it, it seems that the cPickleCache implementation gracefully deals with this situation, whereas the Python implementation raises TypeError---a (gratuitous?) difference.

```C
        while (PyDict_Next(state, &i, &d_key, &d_value)) {
            /* normally the keys for instance attributes are
               interned.  we should try to do that here. */
            if (NATIVE_CHECK_EXACT(d_key)) {
                Py_INCREF(d_key);
                INTERN_INPLACE(&d_key);
                Py_DECREF(d_key);
            }
            if (PyObject_SetItem(*dict, d_key, d_value) < 0)
                return NULL;
        }
```

It seems like the Python implementation should gracefully handle it too.